File: taurus-protect-sdk-typescript/tests/unit/mappers/governance-cell-vectors.test.ts
/**
 * Cross-SDK cell wire-format alignment.
 *
 * Every entry in the shared golden-vector file (produced by the Go SDK) must
 * encode to exactly the recorded bytes and decode back to the same typed cell.
 * The same file is consumed by the Go/Java/Python suites, so this pins
 * byte-for-byte parity across all four SDKs.
 */
import * as fs from 'fs';
import * as path from 'path';
import type { RuleCell } from '../../../src/models/rule-cell';
import { ruleCellFromBytes, ruleCellToBytes } from '../../../src/mappers/rule-cell-codec';

const VECTORS_PATH = path.resolve(
  __dirname,
  '../../../../scripts/resources/governance-cell-vectors.json'
);

interface CellVector {
  description: string;
  column_type: string;
  cell_type: string;
  typed_value_json: unknown;
  wire_base64: string;
}

const RAW_DESCRIPTION = 'raw cell (unknown cell type preserved verbatim)';

// Hardcoded cell cases keyed by the shared vectors' `description`, mirroring the
// Go SDK's allCellCases table. The vectors file is the byte oracle.
const CASES: Record<string, { columnType: string; cell: RuleCell }> = {
  'fiat amount any': { columnType: 'RuleFiatAmount', cell: { kind: 'FiatAmountAny' } },
  'fiat amount is zero': { columnType: 'RuleFiatAmount', cell: { kind: 'FiatAmountIsZero' } },
  'fiat amount range': { columnType: 'RuleFiatAmount', cell: { kind: 'FiatAmountRange', minAmount: '1000', maxAmount: '50000' } },

  'source any': { columnType: 'RuleSource', cell: { kind: 'SourceAny' } },
  'source internal wallet': { columnType: 'RuleSource', cell: { kind: 'SourceInternalWallet', path: "m/44'/60'/0'" } },
  'source internal address': { columnType: 'RuleSource', cell: { kind: 'SourceInternalAddress', address: '0xabc', path: "m/44'/60'/0'/0/0" } },
  'source any exchange': { columnType: 'RuleSource', cell: { kind: 'SourceAnyExchange' } },
  'source exchange': { columnType: 'RuleSource', cell: { kind: 'SourceExchange', label: 'kraken-main' } },
  'source external address': { columnType: 'RuleSource', cell: { kind: 'SourceExternalAddress', address: '0xdef', memo: 'memo-1' } },

  'destination any': { columnType: 'RuleDestination', cell: { kind: 'DestinationAny' } },
  'destination internal wallet': { columnType: 'RuleDestination', cell: { kind: 'DestinationInternalWallet', path: "m/44'/60'/1'" } },
  'destination internal address': { columnType: 'RuleDestination', cell: { kind: 'DestinationInternalAddress', address: '0x111', path: "m/44'/60'/1'/0/0" } },
  'destination external address': { columnType: 'RuleDestination', cell: { kind: 'DestinationExternalAddress', address: '0x222', memo: 'dest-memo' } },
  'destination any exchange': { columnType: 'RuleDestination', cell: { kind: 'DestinationAnyExchange' } },
  'destination exchange': { columnType: 'RuleDestination', cell: { kind: 'DestinationExchange', label: 'binance-desk', memo: 'x' } },
  'destination contract address': { columnType: 'RuleDestination', cell: { kind: 'DestinationContractAddress', address: '0x333', name: 'USDC', symbol: 'USDC', blockchain: 'ETH' } },
  'destination contract address (unknown blockchain)': { columnType: 'RuleDestination', cell: { kind: 'DestinationContractAddress', address: '0x1', name: '', symbol: '', blockchain: '4242' } },
  'destination any external address': { columnType: 'RuleDestination', cell: { kind: 'DestinationAnyExternalAddress' } },
  'destination any contract address': { columnType: 'RuleDestination', cell: { kind: 'DestinationAnyContractAddress' } },

  'string equal any': { columnType: 'RuleStringEqual', cell: { kind: 'StringEqualAny' } },
  'string equal empty': { columnType: 'RuleStringEqual', cell: { kind: 'StringEqualEmpty' } },
  'string equal value': { columnType: 'RuleStringEqual', cell: { kind: 'StringEqualValue', value: 'contract-id-42' } },

  'bytes equal any': { columnType: 'RuleBytesEqual', cell: { kind: 'BytesEqualAny' } },
  'bytes equal empty': { columnType: 'RuleBytesEqual', cell: { kind: 'BytesEqualEmpty' } },
  'bytes equal value': { columnType: 'RuleBytesEqual', cell: { kind: 'BytesEqualValue', value: new Uint8Array([0xde, 0xad, 0xbe, 0xef]) } },

  'string array equal any': { columnType: 'RuleStringArrayEqual', cell: { kind: 'StringArrayEqualAny' } },
  'string array equal empty': { columnType: 'RuleStringArrayEqual', cell: { kind: 'StringArrayEqualEmpty' } },
  'string array equal value': { columnType: 'RuleStringArrayEqual', cell: { kind: 'StringArrayEqualValue', values: ['a', 'b', 'c'] } },

  'integer greater any': { columnType: 'RuleIntegerGreater', cell: { kind: 'IntegerGreaterAny' } },
  'integer greater positive value': { columnType: 'RuleIntegerGreater', cell: { kind: 'IntegerGreaterValue', value: 50n } },
  'integer greater negative value': { columnType: 'RuleIntegerGreater', cell: { kind: 'IntegerGreaterValue', value: -50n } },
  'integer greater zero': { columnType: 'RuleIntegerGreater', cell: { kind: 'IntegerGreaterValue', value: 0n } },

  'uinteger greater any': { columnType: 'RuleUIntegerGreater', cell: { kind: 'UIntegerGreaterAny' } },
  'uinteger greater is zero': { columnType: 'RuleUIntegerGreater', cell: { kind: 'UIntegerGreaterIsZero' } },
  'uinteger greater value': { columnType: 'RuleUIntegerGreater', cell: { kind: 'UIntegerGreaterValue', value: 18446744073709551615n } },
  'uinteger greater is equal': { columnType: 'RuleUIntegerGreater', cell: { kind: 'UIntegerGreaterIsEqual', value: 10n } },

  'whitelisted contract any': { columnType: 'RuleWhitelistedContract', cell: { kind: 'WhitelistedContractAny' } },
  'whitelisted contract address': { columnType: 'RuleWhitelistedContract', cell: { kind: 'WhitelistedContractAddress', address: '0x444', name: 'DAI', symbol: 'DAI', blockchain: 'ETH' } },
};

function toB64(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString('base64');
}
function fromB64(s: string): Uint8Array {
  return new Uint8Array(Buffer.from(s, 'base64'));
}

describe('governance cell golden vectors (cross-SDK parity)', () => {
  const vectors: CellVector[] = JSON.parse(fs.readFileSync(VECTORS_PATH, 'utf-8'));

  it('covers every hardcoded case (grammar completeness)', () => {
    const described = new Set(vectors.map((v) => v.description));
    for (const desc of Object.keys(CASES)) {
      expect(described.has(desc)).toBe(true);
    }
    // vectors = the typed cases + one raw case
    expect(vectors.length).toBe(Object.keys(CASES).length + 1);
  });

  for (const v of vectorsFor()) {
    it(`encodes + decodes: ${v.description}`, () => {
      if (v.description === RAW_DESCRIPTION) {
        const wire = fromB64(v.wire_base64);
        const decoded = ruleCellFromBytes(v.column_type, wire);
        expect(decoded.kind).toBe('RawCell');
        expect(toB64(ruleCellToBytes(v.column_type, decoded))).toBe(v.wire_base64);
        return;
      }
      const c = CASES[v.description];
      expect(c).toBeDefined();
      expect(c.columnType).toBe(v.column_type);
      // encode(typed) === recorded wire bytes (cross-SDK byte parity)
      expect(toB64(ruleCellToBytes(c.columnType, c.cell))).toBe(v.wire_base64);
      // decode(wire) === typed
      expect(ruleCellFromBytes(c.columnType, fromB64(v.wire_base64))).toEqual(c.cell);
    });
  }

  // Read vectors once at module load for the generated `it` blocks.
  function vectorsFor(): CellVector[] {
    return JSON.parse(fs.readFileSync(VECTORS_PATH, 'utf-8'));
  }
});

File: taurus-protect-sdk-java/client/src/test/java/com/taurushq/sdk/protect/client/mapper/GovernanceCellVectorsTest.java
package com.taurushq.sdk.protect.client.mapper;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.taurushq.sdk.protect.client.model.rulescontainer.RuleCell;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

/**
 * Cross-SDK cell wire-format alignment.
 *
 * <p>Every entry in the shared golden-vector file (produced by the Go SDK) must
 * encode to exactly the recorded bytes and decode back to the same typed cell.
 * The same file is consumed by the Go/Python/TypeScript suites, pinning
 * byte-for-byte parity across all four SDKs.
 */
class GovernanceCellVectorsTest {

    private static final String RAW_DESCRIPTION = "raw cell (unknown cell type preserved verbatim)";

    private static JsonArray vectors;

    /**
     * Typed cell cases keyed by the shared vectors' {@code description}, mirroring
     * the Go SDK's allCellCases table. The vectors file is the byte oracle; the
     * expected column type is derived from the cell itself via
     * {@link RuleCellCodec#cellFamily}.
     */
    private static Map<String, RuleCell> cases() {
        Map<String, RuleCell> m = new LinkedHashMap<>();
        m.put("fiat amount any", new RuleCell.FiatAmountAny());
        m.put("fiat amount is zero", new RuleCell.FiatAmountIsZero());
        m.put("fiat amount range", new RuleCell.FiatAmountRange("1000", "50000"));
        m.put("source any", new RuleCell.SourceAny());
        m.put("source internal wallet", new RuleCell.SourceInternalWallet("m/44'/60'/0'"));
        m.put("source internal address", new RuleCell.SourceInternalAddress("0xabc", "m/44'/60'/0'/0/0"));
        m.put("source any exchange", new RuleCell.SourceAnyExchange());
        m.put("source exchange", new RuleCell.SourceExchange("kraken-main"));
        m.put("source external address", new RuleCell.SourceExternalAddress("0xdef", "memo-1"));
        m.put("destination any", new RuleCell.DestinationAny());
        m.put("destination internal wallet", new RuleCell.DestinationInternalWallet("m/44'/60'/1'"));
        m.put("destination internal address", new RuleCell.DestinationInternalAddress("0x111", "m/44'/60'/1'/0/0"));
        m.put("destination external address", new RuleCell.DestinationExternalAddress("0x222", "dest-memo"));
        m.put("destination any exchange", new RuleCell.DestinationAnyExchange());
        m.put("destination exchange", new RuleCell.DestinationExchange("binance-desk", "x"));
        m.put("destination contract address", new RuleCell.DestinationContractAddress("0x333", "USDC", "USDC", "ETH"));
        m.put("destination contract address (unknown blockchain)", new RuleCell.DestinationContractAddress("0x1", "", "", "4242"));
        m.put("destination any external address", new RuleCell.DestinationAnyExternalAddress());
        m.put("destination any contract address", new RuleCell.DestinationAnyContractAddress());
        m.put("string equal any", new RuleCell.StringEqualAny());
        m.put("string equal empty", new RuleCell.StringEqualEmpty());
        m.put("string equal value", new RuleCell.StringEqualValue("contract-id-42"));
        m.put("bytes equal any", new RuleCell.BytesEqualAny());
        m.put("bytes equal empty", new RuleCell.BytesEqualEmpty());
        m.put("bytes equal value", new RuleCell.BytesEqualValue(new byte[] {(byte) 0xDE, (byte) 0xAD, (byte) 0xBE, (byte) 0xEF}));
        m.put("string array equal any", new RuleCell.StringArrayEqualAny());
        m.put("string array equal empty", new RuleCell.StringArrayEqualEmpty());
        m.put("string array equal value", new RuleCell.StringArrayEqualValue(Arrays.asList("a", "b", "c")));
        m.put("integer greater any", new RuleCell.IntegerGreaterAny());
        m.put("integer greater positive value", new RuleCell.IntegerGreaterValue(BigInteger.valueOf(50)));
        m.put("integer greater negative value", new RuleCell.IntegerGreaterValue(BigInteger.valueOf(-50)));
        m.put("integer greater zero", new RuleCell.IntegerGreaterValue(BigInteger.ZERO));
        m.put("uinteger greater any", new RuleCell.UIntegerGreaterAny());
        m.put("uinteger greater is zero", new RuleCell.UIntegerGreaterIsZero());
        m.put("uinteger greater value", new RuleCell.UIntegerGreaterValue(new BigInteger("18446744073709551615")));
        m.put("uinteger greater is equal", new RuleCell.UIntegerGreaterIsEqual(BigInteger.valueOf(10)));
        m.put("whitelisted contract any", new RuleCell.WhitelistedContractAny());
        m.put("whitelisted contract address", new RuleCell.WhitelistedContractAddress("0x444", "DAI", "DAI", "ETH"));
        return m;
    }

    @BeforeAll
    static void loadVectors() throws IOException {
        String[] candidates = {
                "../../scripts/resources/governance-cell-vectors.json", // from client/
                "../scripts/resources/governance-cell-vectors.json",    // from sdk-java/
                "scripts/resources/governance-cell-vectors.json",       // from repo root
        };
        for (String candidate : candidates) {
            Path p = Paths.get(candidate).toAbsolutePath().normalize();
            if (Files.exists(p)) {
                String json = new String(Files.readAllBytes(p), StandardCharsets.UTF_8);
                vectors = JsonParser.parseString(json).getAsJsonArray();
                return;
            }
        }
        throw new IOException("Cannot find governance-cell-vectors.json. Working directory: " + System.getProperty("user.dir"));
    }

    @Test
    void vectorsCoverAllCases() {
        Map<String, RuleCell> cases = cases();
        for (String desc : cases.keySet()) {
            boolean found = false;
            for (JsonElement e : vectors) {
                if (desc.equals(e.getAsJsonObject().get("description").getAsString())) {
                    found = true;
                    break;
                }
            }
            assertTrue(found, "vector missing for " + desc);
        }
        assertEquals(cases.size() + 1, vectors.size(), "typed cases + one raw case");
    }

    @Test
    void cellVectorsEncodeAndDecode() {
        Map<String, RuleCell> cases = cases();
        for (JsonElement e : vectors) {
            JsonObject v = e.getAsJsonObject();
            String desc = v.get("description").getAsString();
            String colType = v.get("column_type").getAsString();
            byte[] wire = Base64.getDecoder().decode(v.get("wire_base64").getAsString());

            if (RAW_DESCRIPTION.equals(desc)) {
                RuleCell decoded = RuleCellCodec.decode(colType, wire);
                assertTrue(decoded instanceof RuleCell.RawCell, desc + ": expected RawCell, got " + decoded);
                assertArrayEquals(wire, RuleCellCodec.encode(colType, decoded), desc);
                continue;
            }

            RuleCell cell = cases.get(desc);
            assertTrue(cell != null, "no case for vector " + desc);
            assertEquals(colType, RuleCellCodec.cellFamily(cell), desc);
            // encode(typed) == recorded wire bytes (cross-SDK byte parity)
            assertArrayEquals(wire, RuleCellCodec.encode(colType, cell), desc);
            // decode(wire) == typed
            assertEquals(cell, RuleCellCodec.decode(colType, wire), desc);
        }
    }
}
